### PR TITLE
C++: Fix indirect taint

### PR DESCRIPTION
**Pull request by MathiasVP**
The motivation for this PR was that we failed to track this flow when the source is whatever `indirect_source()` points to:
```cpp
void test_indirect_taint() {
  std::map<int, int*> m;
  int* p = indirect_source();
  m[1] = p;
  int* q = m[1];
  sink(q); // $ MISSING: ast ir
}
```

The fix was in our interpretation of models (in `ModelUtil.qll`) so that the model for `operator[]` correctly carried indirect flow from `*p` to `m` (this causes pointer/pointee conflation but since these are _taint_ models that's fine).

Finally, fixing the interpretation of models revealed a missing flow step for smart pointers. Specifically, we didn't model that `operator->` and `get` performed a load.

### CODE REVIEW

**Comment by MathiasVP** [comment]
@jketema I've assigned you as a reviewer since it fixes a problem you've observed. Let me know if you need any assistance on the reviewing side 🙇!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll [hunk 1]
```diff
@@ -32,25 +32,34 @@ DataFlow::Node callInput(CallInstruction call, FunctionInput input) {
...
   or
   // The side effect of a call on the value pointed to by an argument or qualifier
-  exists(int index, int indirectionIndex |
+  exists(int index |
     result.(IndirectArgumentOutNode).getArgumentIndex() = index and
-    result.(IndirectArgumentOutNode).getIndirectionIndex() = indirectionIndex and
+    result.(IndirectArgumentOutNode).getIndirectionIndex() = indirectionIndex - 1 and
     result.(IndirectArgumentOutNode).getCallInstruction() = call and
-    output.isParameterDerefOrQualifierObject(index, indirectionIndex)
+    output.isParameterDerefOrQualifierObject(index, indirectionIndex - 1)
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:51]: Why are the subtractions needed here?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll [hunk 2]
```diff
@@ -76,19 +85,16 @@ private IndirectReturnOutNode getIndirectReturnOutNode(CallInstruction call, int
...
 Node callOutput(CallInstruction call, FunctionOutput output, int d) {
-  exists(DataFlow::Node n | n = callOutput(call, output) and d > 0 |
+  exists(DataFlow::Node n, int indirectionIndex |
+    n = callOutputWithIndirectionIndex(call, output, indirectionIndex) and d > 0
+  |
     // The return value
-    result = getIndirectReturnOutNode(n.asInstruction(), d)
+    result = callOutputWithIndirectionIndex(call, output, indirectionIndex + d)
     or
+    n = callOutputWithIndirectionIndex(call, output, indirectionIndex) and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:94]: Why is this line needed? It just seems to repeat the first half of line 89.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll [hunk 3]
```diff
@@ -76,19 +85,16 @@ private IndirectReturnOutNode getIndirectReturnOutNode(CallInstruction call, int
...
     // we conflate this with the underlying `CallInstruction`.
-    not exists(getIndirectReturnOutNode(call, d)) and
+    not exists(getIndirectReturnOutNode(call, indirectionIndex + d)) and
     n = result
-    or
-    // The side effect of a call on the value pointed to by an argument or qualifier
-    exists(Operand operand, int indirectionIndex |
-      Ssa::outNodeHasAddressAndIndex(n, operand, indirectionIndex) and
-      Ssa::outNodeHasAddressAndIndex(result, operand, indirectionIndex + d)
-    )
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:92]: Why is this no longer needed?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:94]: Oops. You're right. This was a leftover of my quick-eval debugging.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll]: Fixed in 1fce26534f7c92ca349038889f7b8c9fd96c6bf6.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll (hunk 3 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:92]: This is now handled by the `result = callOutputWithIndirectionIndex(call, output, indirectionIndex + d)` line because `callOutputWithIndirectionIndex` has a case for `IndirectArgumentOutNode`. So when we have:
```ql
n = callOutputWithIndirectionIndex(call, output, indirectionIndex) and
result = callOutputWithIndirectionIndex(call, output, indirectionIndex + d)
```
in `callOutput` we get an equivalent condition.

Does that make sense?

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:92]: Partially. My main problem is, that it's rather hard for me to map the old code on the new code, because it uses different predicates.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ModelUtil.qll:92]: Yeah, it's not an easy commit to review as it's not a clear line-for-line mapping 😢. The main point of the commit is to get an indirection index out of `callOutput/2` (which has been renamed to `callOutputWithIndirectionIndex`) so that we can compute `indirectionIndex + d` on line 92 to "increment" the number of indirections of the resulting node by `d`.

**Review by jketema — APPROVED** [review]
